### PR TITLE
Add CLI back

### PR DESCRIPTION
**Pull request by rany2**
This is not completely done, some commands weren't added

### CODE REVIEW

**Comment by rany2** [comment]
I need help with `components` and `savedrooms`.   

`components` usually works fine but I couldn't support any kind of lock because the event is emitted for every component (can't figure out when I reached the end of the batch to know if I should lock).

`savedrooms` doesn't seem possible to me but if it is and i missed how to add support please tell me

**Comment by schwatter** [comment]
Great progress.

  charge         Returns to charger <-- works
  clean          Auto clean  <-- works
  components     Get robot components life span  <-- works
  createconfig   logs in with specified email; run this first <-- it started, but i stopped because i have already a running setup
  customarea     Cleans provided area(s), ex: "-602,1812,800,723" <-- not tested
  exportlivemap  Get robot map and save it [filepath ex:...  <-- works
  getcleanlogs   Get Clean Logs <-- never ending
  getrooms       Get saved rooms <-- not supportet
  pause          Pause the robot  <-- works
  playsound      Play welcome sound  <-- works
  resume         Resume the robot  <-- works
  setfanspeed    Set Clean Speed <-- works
  setwaterlevel  Set Water Level  <-- works
  spotarea       Cleans provided rooms(s), ex: "0,1" | Use GetRooms to see... <-- not tested because of not knowing the rooms
  stats          Get stats <-- works
  statuses       Get robot statuses [Status,Battery,FanSpeed,WaterLevel] <-- works

Greeting schwatter

**Comment by edenhaus** [comment]
> `components` usually works fine but I couldn't support any kind of lock because the event is emitted for every component (can't figure out when I reached the end of the batch to know if I should release lock).

I can modify it, that all components are send out in one event, but I will do that in a later moment.

**Comment by rany2** [comment]
@schwatter 
>  getcleanlogs Get Clean Logs <-- never ending

It is never ending because we need https://github.com/And3rsL/Deebotozmo/pull/54 to be merged as well

**Comment by rany2** [comment]
@edenhaus thanks, any tips on the savedrooms issue?

File: deebotozmo/cli.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+    await vacbot.run()
+
+    lock = asyncio.Lock()
+    await lock.acquire()
+    async def on_map(event: MapEvent):
+        with open(filepath, "wb") as fh:
+            fh.write(base64.decodebytes(vacbot.bot.map.get_base64_map()))
+        lock.release()
+    listener = vacbot.bot.map.mapEvents.subscribe(on_map)
+    await vacbot.bot.execute_command(GetMapTrace())
```
Comment (edenhaus) [inline deebotozmo/cli.py:303]:
```suggestion
```
see above, also remove the next two lines

File: deebotozmo/cli.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+        print(str(event.type) + ": " + str(event.percent) + "%")
+    listener = vacbot.bot.lifespanEvents.subscribe(on_lifespan_event)
+    await vacbot.bot.execute_command(GetLifeSpan())
+    vacbot.bot.lifespanEvents.unsubscribe(listener)
+    await vacbot.goodbye()
+
+@cli.command(help='Get saved rooms')
+@coro
+async def getrooms():
+    print ("Not supported, no clue how to make it work.")
```
Comment (edenhaus) [inline deebotozmo/cli.py:281]: Use `listener = vacbot.bot.map.roomsEvents.subscribe(on_rooms)`
Similar to the other events

File: deebotozmo/cli.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    listener = vacbot.bot.map.mapEvents.subscribe(on_map)
+    await vacbot.bot.execute_command(GetMapTrace())
+    await vacbot.bot.execute_command(GetPos())
+    await vacbot.bot.execute_command(GetMajorMap())
+    await lock.acquire()
+    vacbot.bot.map.mapEvents.unsubscribe(listener)
+
+    await vacbot.goodbye()
+
+class dologin:
```
Comment (edenhaus) [inline deebotozmo/cli.py:311]: Use a better name and in Python a class start uppercase

File: deebotozmo/cli.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+                                    verify_ssl=bool(config['verify_ssl']))
+        await self.api.login()
+
+        self.devices_ = await self.api.get_devices()
+
+        self.auth = await self.api.get_request_auth()
+        self.bot = VacuumBot(self.session, self.auth, self.devices_[0], continent=config['continent'],
+                                   country=config['country'], verify_ssl=bool(config['verify_ssl']))
+
+        self.mqtt = EcovacsMqtt(self.auth, continent=config['continent'], country=config['country'])
```
Comment (edenhaus) [inline deebotozmo/cli.py:335]:
```suggestion
```
You don't need it as mqtt is used to get events, which are triggered from other instances. As you unscribe straight away you would never get these events. I don't know how you can run mqtt in the background and inform you about new events

File: deebotozmo/cli.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+import logging
+import os
+import platform
+import sys
+import time
+from dataclasses import asdict
+from functools import wraps
+
+import aiohttp
+import click
```
Comment (edenhaus) [inline deebotozmo/cli.py:17]: `ModuleNotFoundError: No module named 'click'`

Please add this requirement to setup.py

**Review by edenhaus — CHANGES_REQUESTED** [review]
I quickly looked at your code on high level. A functional test was already done by @schwatter

**Comment by rany2** [comment]
@edenhaus I don't have any rooms so getrooms ends up hanging indefinitely, is there any way I could detect if we have no rooms?

**Comment by edenhaus** [comment]
> @edenhaus I don't have any rooms so getrooms ends up hanging indefinitely, is there any way I could detect if we have no rooms?

Sorry I have no idea.

File: deebotozmo/cli.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+    await vacbot.run()
+
+    lock = asyncio.Lock()
+    await lock.acquire()
+    async def on_clean_event(event: CleanLogEvent):
+        print(json.dumps(asdict(event)))
+        lock.release()
+
+    listener = vacbot.bot.cleanLogsEvents.subscribe(on_clean_event)
+    vacbot.bot.cleanLogsEvents.request_refresh()
```
Comment (edenhaus) [inline deebotozmo/cli.py:198]:
```suggestion
```

Please remove the code line as it is not required and with this line you will fetch the data **twice**!!

Please also remove the other `request_refresh` code lines as I requested in the first review

File: deebotozmo/cli.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+from deebotozmo.util import md5
+from deebotozmo.vacuum_bot import VacuumBot
+
+_LOGGER = logging.getLogger(__name__)
+
+def coro(f):
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+        loop = asyncio.get_event_loop()
+        return loop.run_until_complete(f(*args, **kwargs))
```
[the remark this answers is not in the document]
Comment (rany2) [reply deebotozmo/cli.py:35]: I don't understand what you mean

**Review by edenhaus — CHANGES_REQUESTED** [review]
Please read my suggestion carefully, as I requested this exact problem already in my first review

File: deebotozmo/cli.py (hunk 6 above)
Comment (rany2) [reply deebotozmo/cli.py:198]: I had instances where I missed these and never got any data, I'll try again

File: deebotozmo/cli.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (edenhaus) [reply deebotozmo/cli.py:35]: You call in many methods these three lines...
Example:
```python
cli.command(help='Play welcome sound')
@coro
async def playsound():
    vacbot = DoLogin()
    await vacbot.run()
    await vacbot.bot.execute_command(PlaySound())
    await vacbot.goodbye()

@cli.command(help='Auto clean')
@coro
async def clean():
    vacbot = DoLogin()
    await vacbot.run()
    await vacbot.bot.execute_command(CleanStart())
    await vacbot.goodbye()
```

playsound and clean are the same except one line.
I would suggest something like this (not tested):

```python
@coro
def with_login(f):
    @wraps(f)
    def wrapper(*args, **kwargs):
           vacbot = DoLogin()
           await vacbot.run()
           await f(*args, **kwargs)
           await vacbot.goodbye()
   return wrapper

@cli.command(help='Play welcome sound')
@with_login
async def playsound():
    await vacbot.bot.execute_command(PlaySound())

@cli.command(help='Auto clean')
@with_login
async def clean():
    await vacbot.bot.execute_command(CleanStart())
```

[the remark this answers is not in the document]
Comment (rany2) [reply deebotozmo/cli.py:35]: I see, I thought you meant that you wanted them to become global variables. Will do

[the remark this answers is not in the document]
Comment (rany2) [reply deebotozmo/cli.py:35]: Is something like https://github.com/And3rsL/Deebotozmo/pull/55/commits/47eed02b9e62ac8d7c29cff0b28621ec0d707720 acceptable?

**Comment by schwatter** [comment]
getrooms tested

```
root@raspi4fhem:~# python3 /usr/local/lib/python3.7/dist-packages/deebotozmo/cli.py getrooms
5 Dining Room
6 Corridor
7 Bathroom
8 Living Room
root@raspi4fhem:~#
```

What was the right syntax for cleaning say two rooms with spotarea. I think now it is a bit different?

**Comment by rany2** [comment]
@schwatter Does it work now?

**Comment by schwatter** [comment]
Ok check, it work now. 

So the robot is clean. Now when i try to stop and send charge i get now keyerror.

```
root@raspi4fhem:~# python3 /usr/local/lib/python3.7/dist-packages/deebotozmo/cli.py spotarea 5,8 1
root@raspi4fhem:~# python3 /usr/local/lib/python3.7/dist-packages/deebotozmo/cli.py charge
--- Logging error ---
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/deebotozmo/ecovacs_mqtt.py", line 48, in _on_message
    await bot.handle(topic_split[2], data, False)
  File "/usr/local/lib/python3.7/dist-packages/deebotozmo/vacuum_bot.py", line 191, in handle
    await self._map.handle(event_name, event_data, requested)
  File "/usr/local/lib/python3.7/dist-packages/deebotozmo/map.py", line 117, in handle
    self._handle_map_sub_set(event_data)
  File "/usr/local/lib/python3.7/dist-packages/deebotozmo/map.py", line 172, in _handle_map_sub_set
    subtype = int(event_data["subtype"])
KeyError: 'subtype'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.7/logging/__init__.py", line 1034, in emit
    msg = self.format(record)
  File "/usr/lib/python3.7/logging/__init__.py", line 880, in format
    return fmt.format(record)
  File "/usr/lib/python3.7/logging/__init__.py", line 619, in format
    record.message = record.getMessage()
  File "/usr/lib/python3.7/logging/__init__.py", line 380, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "/usr/local/lib/python3.7/dist-packages/deebotozmo/cli.py", line 314, in <module>
    cli()
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1289, in invoke
    rv.append(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/deebotozmo/cli.py", line 36, in wrapper
    return loop.run_until_complete(f(*args, **kwargs))
  File "/usr/lib/python3.7/asyncio/base_events.py", line 571, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.7/asyncio/base_events.py", line 539, in run_forever
    self._run_once()
  File "/usr/lib/python3.7/asyncio/base_events.py", line 1775, in _run_once
    handle._run()
  File "/usr/lib/python3.7/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.7/dist-packages/deebotozmo/ecovacs_mqtt.py", line 51, in _on_message
    _LOGGER.error("An exception occurred", err, exc_info=True)
Message: 'An exception occurred'
Arguments: (KeyError('subtype'),)
root@raspi4fhem:~#
```

**Comment by rany2** [comment]
@schwatter I don't think it's a bug with the CLI, maybe it was a one off and rerunning it would work fine?

**Comment by edenhaus** [comment]
Without logs I cannot identify the problem, but it has nothing to do with the cli. If it happen again please open a new issue.

**Comment by edenhaus** [comment]
Are one or both of you willing to maintain they cli and test it?
I ask this question as I currently do some refactoring on the events and I don't have the time to maintain the cli and test it for every change

**Comment by rany2** [comment]
@edenhaus I'm willing to continue maintaining and testing the CLI

**Comment by schwatter** [comment]
Hallo,

finally i'm on python3.10 and testing the library with cli version 3.0.1

charge         Returns to charger  <-- works
clean          Auto clean <--works
components     Get robot components life span <--works
createconfig   logs in with specified email; run this first <-- not testes yet because of my old working config
customarea     Cleans provided area(s), ex: "-602,1812,800,723" <-- not tested
exportlivemap  Get robot map and save it [filepath ex:... <-- at the moment some loop
getcleanlogs   Get Clean Logs <-- works
getdevices     Get Devices <-- i have just one 950, but it show up. So works
getrooms       Get saved rooms <-- runtime error

```
root@raspi4fhem:~# deebotozmo getrooms
asyncio    ERROR    Task exception was never retrieved
future: <Task finished name='Task-10' coro=<EventEmitter._call_refresh_function() done, defined at /usr/local/lib/python3.10/site-packages/deebotozmo/event_emitter.py:97> exception=RuntimeError('Session is closed')>
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/deebotozmo/event_emitter.py", line 104, in _call_refresh_function
    await self._refresh_function()
  File "/usr/local/lib/python3.10/site-packages/deebotozmo/util.py", line 42, in refresh
    await execute_command(commands[0])
  File "/usr/local/lib/python3.10/site-packages/deebotozmo/vacuum_bot.py", line 147, in execute_command
    response = await self.json.send_command(command, self.vacuum)
  File "/usr/local/lib/python3.10/site-packages/deebotozmo/ecovacs_json.py", line 47, in send_command
    async with self._session.post(
  File "/usr/local/lib/python3.10/site-packages/aiohttp/client.py", line 1117, in __aenter__
    self._resp = await self._coro
  File "/usr/local/lib/python3.10/site-packages/aiohttp/client.py", line 381, in _request
    raise RuntimeError("Session is closed")
RuntimeError: Session is closed
root@raspi4fhem:~#
```
pause          Pause the robot <-- works
playsound      Play welcome sound <-- works
resume         Resume the robot <-- works
setfanspeed    Set Clean Speed <-- works
setwaterlevel  Set Water Level <-- works
spotarea       Cleans provided rooms(s), ex: "0,1" | Use GetRooms to... <-- Error: Missing argument 'CLEANINGS'.

```
root@raspi4fhem:~# deebotozmo spotarea "0,4"
Usage: deebotozmo spotarea [OPTIONS] ROOMS CLEANINGS
Try 'deebotozmo spotarea --help' for help.

Error: Missing argument 'CLEANINGS'.
root@raspi4fhem:~#
```
stats          Get stats <-- works
statuses       Get robot statuses [Status,Battery,FanSpeed,WaterLevel] <-- works

**Comment by schwatter** [comment]
Ok, but still problems. Just one area or two make problems.

```
root@raspi4fhem:~# deebotozmo spotarea 0
Usage: deebotozmo spotarea [OPTIONS] ROOMS CLEANINGS
Try 'deebotozmo spotarea --help' for help.

Error: Missing argument 'CLEANINGS'.
root@raspi4fhem:~# deebotozmo spotarea 0 4
deebotozmo.commands.base WARNING  Command "clean" was not successfully. body={'code': 20011, 'msg': 'fail'}
root@raspi4fhem:~#
```

**Comment by schwatter** [comment]
And another fix for getrooms

```
        event.set()
        listener = util.bot.map.events.rooms.subscribe(on_rooms)
```

change to

```
        event.clear()
        listener = util.bot.map.events.rooms.subscribe(on_rooms)
```

**Pull request merged**